File: php/test/EmailDomainDetailEntityTest.php
<?php
declare(strict_types=1);

// EmailDomainDetail entity test

require_once __DIR__ . '/../lmemail_sdk.php';
require_once __DIR__ . '/Runner.php';

use PHPUnit\Framework\TestCase;
use Voxgig\Struct\Struct as Vs;

class EmailDomainDetailEntityTest extends TestCase
{
    public function test_create_instance(): void
    {
        $testsdk = LmEmailSDK::test(null, null);
        $ent = $testsdk->EmailDomainDetail(null);
        $this->assertNotNull($ent);
    }

    public function test_basic_flow(): void
    {
        $setup = email_domain_detail_basic_setup(null);
        // Per-op sdk-test-control.json skip.
        $_live = !empty($setup["live"]);
        foreach (["load"] as $_op) {
            [$_shouldSkip, $_reason] = Runner::is_control_skipped("entityOp", "email_domain_detail." . $_op, $_live ? "live" : "unit");
            if ($_shouldSkip) {
                $this->markTestSkipped($_reason ?? "skipped via sdk-test-control.json");
                return;
            }
        }
        // The basic flow consumes synthetic IDs from the fixture. In live mode
        // without an *_ENTID env override, those IDs hit the live API and 4xx.
        if (!empty($setup["synthetic_only"])) {
            $this->markTestSkipped("live entity test uses synthetic IDs from fixture — set LMEMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID JSON to run live");
            return;
        }
        $client = $setup["client"];

        // Bootstrap entity data from existing test data.
        $email_domain_detail_ref01_data_raw = Vs::items(Helpers::to_map(
            Vs::getpath($setup["data"], "existing.email_domain_detail")));
        $email_domain_detail_ref01_data = null;
        if (count($email_domain_detail_ref01_data_raw) > 0) {
            $email_domain_detail_ref01_data = Helpers::to_map($email_domain_detail_ref01_data_raw[0][1]);
        }

        // LOAD
        $email_domain_detail_ref01_ent = $client->EmailDomainDetail(null);
        $email_domain_detail_ref01_match_dt0 = [
            "id" => $email_domain_detail_ref01_data["id"],
        ];
        $email_domain_detail_ref01_data_dt0_loaded = $email_domain_detail_ref01_ent->load($email_domain_detail_ref01_match_dt0, null);
        $email_domain_detail_ref01_data_dt0_load_result = Helpers::to_map($email_domain_detail_ref01_data_dt0_loaded);
        $this->assertNotNull($email_domain_detail_ref01_data_dt0_load_result);
        $this->assertEquals($email_domain_detail_ref01_data_dt0_load_result["id"], $email_domain_detail_ref01_data["id"]);

    }
}

function email_domain_detail_basic_setup($extra)
{
    Runner::load_env_local();

    $entity_data_file = __DIR__ . '/../../.sdk/test/entity/email_domain_detail/EmailDomainDetailTestData.json';
    $entity_data_source = file_get_contents($entity_data_file);
    $entity_data = json_decode($entity_data_source, true);

    $options = [];
    $options["entity"] = $entity_data["existing"];

    $client = LmEmailSDK::test($options, $extra);

    // Generate idmap.
    $idmap = [];
    foreach (["email_domain_detail01", "email_domain_detail02", "email_domain_detail03"] as $k) {
        $idmap[$k] = strtoupper($k);
    }

    // Detect ENTID env override before envOverride consumes it. When live
    // mode is on without a real override, the basic test runs against synthetic
    // IDs from the fixture and 4xx's. Surface this so the test can skip.
    $entid_env_raw = getenv("LMEMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID");
    $idmap_overridden = $entid_env_raw !== false && str_starts_with(trim($entid_env_raw), "{");

    $env = Runner::env_override([
        "LMEMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID" => $idmap,
        "LMEMAIL_TEST_LIVE" => "FALSE",
        "LMEMAIL_TEST_EXPLAIN" => "FALSE",
        "LMEMAIL_APIKEY" => "NONE",
    ]);

    $idmap_resolved = Helpers::to_map(
        $env["LMEMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID"]);
    if ($idmap_resolved === null) {
        $idmap_resolved = Helpers::to_map($idmap);
    }

    if ($env["LMEMAIL_TEST_LIVE"] === "TRUE") {
        $merged_opts = Vs::merge([
            [
                "apikey" => $env["LMEMAIL_APIKEY"],
            ],
            $extra ?? [],
        ]);
        $client = new LmEmailSDK(Helpers::to_map($merged_opts));
    }

    $live = $env["LMEMAIL_TEST_LIVE"] === "TRUE";
    return [
        "client" => $client,
        "data" => $entity_data,
        "idmap" => $idmap_resolved,
        "env" => $env,
        "explain" => $env["LMEMAIL_TEST_EXPLAIN"] === "TRUE",
        "live" => $live,
        "synthetic_only" => $live && !$idmap_overridden,
        "now" => (int)(microtime(true) * 1000),
    ];
}

File: go/test/email_domain_detail_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/lm-email-sdk/go"
	"github.com/voxgig-sdk/lm-email-sdk/go/core"

	vs "github.com/voxgig-sdk/lm-email-sdk/go/utility/struct"
)

func TestEmailDomainDetailEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.EmailDomainDetail(nil)
		if ent == nil {
			t.Fatal("expected non-nil EmailDomainDetailEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := email_domain_detailBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "email_domain_detail." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set LM_EMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		emailDomainDetailRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.email_domain_detail", setup.data)))
		var emailDomainDetailRef01Data map[string]any
		if len(emailDomainDetailRef01DataRaw) > 0 {
			emailDomainDetailRef01Data = core.ToMapAny(emailDomainDetailRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = emailDomainDetailRef01Data

		// LOAD
		emailDomainDetailRef01Ent := client.EmailDomainDetail(nil)
		emailDomainDetailRef01MatchDt0 := map[string]any{
			"id": emailDomainDetailRef01Data["id"],
		}
		emailDomainDetailRef01DataDt0Loaded, err := emailDomainDetailRef01Ent.Load(emailDomainDetailRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		emailDomainDetailRef01DataDt0LoadResult := core.ToMapAny(entityData(emailDomainDetailRef01DataDt0Loaded))
		if emailDomainDetailRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if emailDomainDetailRef01DataDt0LoadResult["id"] != emailDomainDetailRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

	})
}

func email_domain_detailBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "email_domain_detail", "EmailDomainDetailTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read email_domain_detail test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse email_domain_detail test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"email_domain_detail01", "email_domain_detail02", "email_domain_detail03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("LM_EMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"LM_EMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID": idmap,
		"LM_EMAIL_TEST_LIVE":      "FALSE",
		"LM_EMAIL_TEST_EXPLAIN":   "FALSE",
		"LM_EMAIL_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["LM_EMAIL_TEST_EMAIL_DOMAIN_DETAIL_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["LM_EMAIL_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["LM_EMAIL_APIKEY"],
			},
			extra,
		})
		client = sdk.NewLmEmailSDK(core.ToMapAny(mergedOpts))
	}

	live := env["LM_EMAIL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["LM_EMAIL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
